core/src/consensus/fork_choice.rs

```diff
@@ -179,221 +343,60 @@ pub fn select_vote_and_reset_forks(
-                    reset_bank.as_ref().map(|b| b.slot()),
-                    latest_duplicate_ancestor,
-                );
-                failure_reasons.push(HeaviestForkFailures::FailedSwitchThreshold(
-                    heaviest_bank.slot(),
-                    0, // In this case we never actually performed the switch check, 0 for now
-                    0,
-                ));
-                CandidateVoteAndResetBanks {
-                    candidate_vote_bank: None,
-                    reset_bank,
-                    switch_fork_decision,
-                }
-            }
-            _ => CandidateVoteAndResetBanks {
-                candidate_vote_bank: Some(heaviest_bank),
-                reset_bank: Some(heaviest_bank),
-                switch_fork_decision,
-            },
-        }
-    };
+    let initial_switch_fork_decision: SwitchForkDecision = tower.check_switch_threshold(
+        heaviest_bank.slot(),
+        ancestors,
+        descendants,
+        progress,
+        heaviest_bank.total_epoch_stake(),
+        heaviest_bank
+            .epoch_vote_accounts(heaviest_bank.epoch())
+            .expect("Bank epoch vote accounts must contain entry for the bank's own epoch"),
+        latest_validator_votes_for_frozen_banks,
+        fork_choice,
+    );
 
+    let mut failure_reasons = vec![];
     let CandidateVoteAndResetBanks {
         candidate_vote_bank,
         reset_bank,
         switch_fork_decision,
-    } = candidate_vote_and_reset_banks;
+    } = select_candidate_vote_and_reset_banks(
+        heaviest_bank,
+        heaviest_bank_on_same_voted_fork,
+        progress,
+        tower,
+        &mut failure_reasons,
+        initial_switch_fork_decision,
+    );
 
-    if let Some(candidate_vote_bank) = candidate_vote_bank {
-        // If there's a bank to potentially vote on, then make the remaining
-        // checks
-        let (
-            is_locked_out,
-            vote_thresholds,
-            propagated_stake,
-            is_leader_slot,
-            fork_weight,
-            total_threshold_stake,
-            total_epoch_stake,
-        ) = {
-            let fork_stats = progress.get_fork_stats(candidate_vote_bank.slot()).unwrap();
-            let propagated_stats = &progress
-                .get_propagated_stats(candidate_vote_bank.slot())
-                .unwrap();
-            (
-                fork_stats.is_locked_out,
-                &fork_stats.vote_threshold,
-                propagated_stats.propagated_validators_stake,
-                propagated_stats.is_leader_slot,
-                fork_stats.fork_weight(),
-                fork_stats.total_stake,
-                propagated_stats.total_epoch_stake,
-            )
+    let Some(candidate_vote_bank) = candidate_vote_bank else {
+        // No viable candidate to vote on.
+        return SelectVoteAndResetForkResult {
+            vote_bank: None,
+            reset_bank: reset_bank.cloned(),
```

Comment:
Just noting that you combined the two else cases here, but I believe this is correct, no need to case on whether `reset_bank` is `Some`

Reply:
Good eye. Yeah, I was thinking we didn't need two different else legs. We can just clone the None